### PR TITLE
Fix Expr's Eq instance via a newtype wrapper for Doubles

### LINKED ISSUE #1341: `freeIn` incorrectly relies on Expr's Eq instance
This should be `False`:

```
λ> V "x" 0 `freeIn` DoubleLit (0/0)
True
```

https://github.com/dhall-lang/dhall-haskell/blob/08747af8d30a9bcf11490076cdfc8eeca0e6c305/dhall/src/Dhall/Core.hs#L2067-L2075

### PR DESCRIPTION
**Pull request by sjakobi**
Goals:

* We could unskip the `AssertNaN` test: https://github.com/dhall-lang/dhall-haskell/blob/5b136dfa2c42ba41a57aa495091b598cc6414c9f/dhall/tests/Dhall/Test/TypeInference.hs#L47-L52

* Generate NaNs other problematic Doubles in Expr's Arbitrary instance, gaining better test coverage.

* This would also fix #1341.

This doesn't quite compile yet, but I wanted to get some opinions on this approach, before putting in more work.

### CODE REVIEW

File: dhall/src/Dhall/Core.hs [hunk 1]
```diff
@@ -324,6 +327,19 @@ instance Pretty Import where
             Location -> " as Location"
             Code     -> ""
 
+newtype DhallDouble = DhallDouble Double
+    deriving (Show, Data, NFData, Generic)
+
+instance Eq DhallDouble where
+    n == n' =
+        Serialise.serialise (Binary.encode n) ==  Serialise.serialise (Binary.encode n')
```
Comment (sjakobi) [inline dhall/src/Dhall/Core.hs]: If we switch to a definition that doesn't rely on `Dhall.Binary`, we could probably get rid of `Dhall/Binary.hs-boot` altogether.

File: dhall/src/Dhall/Core.hs [hunk 2]
```diff
@@ -324,6 +327,30 @@ instance Pretty Import where
             Location -> " as Location"
             Code     -> ""
 
+-- | This wrapper around 'Prelude.Double' exists for its 'Eq' instance which is
+-- defined via the binary encoding of Dhall @Double@s.
+newtype DhallDouble = DhallDouble { getDhallDouble :: Double }
```
Comment (sjakobi) [inline dhall/src/Dhall/Core.hs]: Any ideas for a better name, anyone!?

File: dhall/src/Dhall/Core.hs [hunk 3]
```diff
@@ -324,6 +325,37 @@ instance Pretty Import where
...
+--
+-- >>> nan = DhallDouble (0/0)
+-- >>> nan == nan
+-- True
+--
+-- This instance is also consistent with with the binary encoding of Dhall @Double@s:
+--
+-- >>> toBytes n = Codec.Serialise.serialise (Dhall.Binary.encode (n :: DhallDouble))
+--
+-- prop> \a b -> (a == b) == (toBytes a == toBytes b)
```
Comment (sjakobi) [inline dhall/src/Dhall/Core.hs]: This is actually not true for `a = -0.0` and `b = 0.0`.

File: dhall/src/Dhall/Core.hs [hunk 4]
```diff
@@ -324,6 +325,37 @@ instance Pretty Import where
...
+instance Eq DhallDouble where
+    DhallDouble a == DhallDouble b
+        | isNaN a && isNaN b = True
+        | otherwise          = a == b
```
Comment (sjakobi) [inline dhall/src/Dhall/Core.hs]: This instance, and using it in `conv`, makes `-0.0` and `0.0` equivalent.

I've opened https://github.com/dhall-lang/dhall-lang/issues/752 to find out whether we actually want this.

Comment (sjakobi) [reply dhall/src/Dhall/Core.hs]: I have updated the instance to treat `-0.0` and `0.0` as different, just as before this patch. 

Depending on the outcome of https://github.com/dhall-lang/dhall-lang/issues/752, we can still change this later.

**Comment by Gabriel439** [comment]
@sjakobi: It looks like `stack-lts-6.yaml` needs an `extra-dep` for `special-values`

**Pull request merged**